Scripts/softclip_graph.py

```diff
@@ -0,0 +1,224 @@
+        elif len(read.positions) == 0:
+            read_data[2] += 1
+        read_data[0] += 1
+
+    return softclipdata, read_data
+
+
+def update_softclipdata(region, softclipdata):
+    """ The update_softclipdata function updates the softclipdata dictionary with new information.
+
+    :param region: A region of a read defined by the cigar string.
+    :param softclipdata: A dictionary with positions and the number of "normal" bases and softclip bases.
+    :return sofclipdata: A dictionary with positions and the number of "normal" bases and softclip bases.
+    """
+    if region[2] == 'normal':
+        index = 0
+    else:
+        index = 1
+
+    for pos in range(region[0], region[1]):
+        if pos in softclipdata:
+            softclipdata[pos][index] += 1
+        else:
+            posdata = [0, 0]
+            posdata[index] += 1
+            softclipdata.update({pos: posdata})
+
+    return softclipdata
+
+
+def softclip_regions(read_start, cigar):
+    """ The softclip regions function iterates over the cigar string and returns the positions and if they are "normal"
+    or a softclipped region.
+
+    :param read_start: An integer representing the start of the read (including unmapped bases)
+    :param cigar: a list containing tuples representing the cigar string.
+    :return regions: a 2d list containing the regions of a read and specifying if they are "normal" or softclipped.
+    """
+    regions = []
+    cursor = read_start
+    for element in cigar:
+        if element[0] == 4:
+            regions.append([cursor, cursor+element[1], 'softclip'])
+        else:
+            regions.append([cursor, cursor+element[1], 'normal'])
+        cursor += element[1]
+
+    return regions
+
+
+def true_start(cigar, matchstart):
+    """ The true_start function receives the cigar string and the starting position of the first match in a read. It
+    returns the start position of the read including unmapped parts.
+
+    :param cigar: a list containing tuples representing the cigar string.
+    :param matchstart: an integer representing the start of the first mapped base of a read.
+    """
+    overshoot = 0
+
+    for element in cigar:
+        if element[0] != 0:
+            overshoot += element[1]
+        else:
+            break
+
+    read_start = matchstart - overshoot
+
+    return read_start
+
+
+def heatmapdata(sofclipdata):
+    """ The heatmapdata function receives the softclipdata and translates it to a more suitable format for a bedfile.
+
+    :param sofclipdata: A dictionary with positions and the number of "normal" bases and softclip bases.
+    :return heatmapdata: A 2d list containing the coordinates and the percentage of sofclipped bases.
+    """
+    heatmapdata = []
+    first = True
+    chromosome = args.region.split(':')[0]
+
+    for pos in sofclipdata:
+        if sofclipdata[pos][1] != 0:
+            ratio = round(sofclipdata[pos][1]/(sofclipdata[pos][0]+sofclipdata[pos][1]), 2)
```

Comment:
Door hier af te ronden creëer je een 'smoothing', is dat de bedoeling?

Reply:
Ja klopt maar het gebeurd om te voorkomen dat er in het barplot nare getallen komen en het zorgt er ook voor dat het makkelijker te compressen in de bed file.